Update documentation for supported token-exchange (#38008)
keycloak/keycloak: docs/documentation/upgrading/topics/changes/changes-26_2_0.adoc

In this release, {project_name} added support for the link:{securing_apps_token_exchange_link}#_standard-token-exchange[Standard token exchange]. In the past {project_name} releases, the {project_name} supported only preview token exchange feature, which is now referred
as link:{securing_apps_token_exchange_link}#_legacy-token-exchange[Legacy token exchange]. In case you used the link:{securing_apps_token_exchange_link}#_internal-token-to-internal-token-exchange[internal-internal token exchange],
it is recommended to migrate to the new standard token exchange instead. Here are some notes, which can help to migrate to the new standard token exchange:

* The feature `token-exchange-standard`, which represents the new Standard token exchange is enabled by default. It is recommended to just
disable the `token-exchange` feature, which represents the Legacy token exchange to make sure that requests will be served by the new standard token exchange.

* It is possible to have both standard and legacy token exchange enabled, which can be useful if you need to cover standard use cases (internal-internal) together with the other token exchange use cases,

> which are supported only by legacy token exchange. Like for instance link:{securing_apps_token_exchange_link}#_external-token-to-internal-token-exchange[external to internal token exchange]. In this case

Reviewer: Does this not imply that legacy token exchange is officially supported?

Author: Yes, good catch! It seems using the world "supported" is tricky, so I've rather replaced it with "implemented" . Also reworked this part a bit (also based on the feedback by @rmartinc ).

{project_name} will serve standard internal to internal requests preferably by the standard token exchange when the other requests will be served by the legacy token exchange. It is detected based on the
parameters if the particular request is standard or not. For instance requests containing non-standard parameters like `requested_issuer` or `requested_subject` are considered legacy.

* Standard token exchange requires to enable switch on the client as described in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-enable[{securing_apps_token_exchange_name}].

* Fine-grained admin permissions are no longer needed and no longer supported for the standard token exchange.

* The most notable change regarding the behaviour of scopes and audiences is, that applied client scopes are based on the client triggering the token exchange request, rather than "target" client specified by the `audience` parameter.
There is support for multiple values of the `audience` parameter as mentioned in the specification. The details are described in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-scope[{securing_apps_token_exchange_name}].

> * Public clients are not allowed anymore to send the token exchange requests. Legacy token exchange supported public clients to exchanging to itself to downscope the original token. This use case can be

Reviewer: Maybe replace "not allowed anymore" with "no longer allowed".

Author: Updated

Read more about this in the https://www.keycloak.org/operator/advanced-configuration[Operator Advanced configuration].

=== Supported standard token exchange

In this release, {project_name} added support for the link:{securing_apps_token_exchange_link}#_standard-token-exchange[Standard token exchange] (Feature `token-exchange-standard:v2`). In the past {project_name} releases,

> the {project_name} had only a preview token exchange feature, which is now referred to as link:{securing_apps_token_exchange_link}#_legacy-token-exchange[Legacy token exchange] (Feature `token-exchange:v1`).

Reviewer 2: ```suggestion
{project_name} had only a preview token exchange feature, which is now referred to as link:{securing_apps_token_exchange_link}#_legacy-token-exchange[Legacy token exchange] (Feature `token-exchange:v1`).
```

> The legacy token exchange still remains preview and it works the same way as in previous releases.  In case you used the link:{securing_apps_token_exchange_link}#_internal-token-to-internal-token-exchange[internal-internal token exchange],

Author: Thanks, updated

> it is recommended to migrate to the new standard token exchange instead.

Reviewer 2: ```suggestion
consider migrating to the new standard token exchange.
```

If needed for some reason, you can stick to the legacy token exchange and it should work the same way as in previous releases. To continue using legacy token exchange, it is not necessary to disable the standard token exchange feature
as your clients will use standard token exchange just if it is explicitly enabled on the {project_name} client as described below. However, migration to the standard token exchange is recommended as that is
officially supported and will be the priority for the {project_name} team in the future.

Here are some notes that can help you migrate to the new standard token exchange:

* The feature `token-exchange-standard`, which represents the new Standard token exchange, is enabled by default. It is recommended to
disable the `token-exchange` feature, which represents the Legacy token exchange, to make sure that requests will be served by the new standard token exchange.

> * It is possible to have both the standard and the legacy token exchange features enabled, which can be useful if you need to cover standard use cases (internal-internal) together with the other token exchange use cases,
> which are implemented only by legacy token exchange. For instance, the link:{securing_apps_token_exchange_link}#_external-token-to-internal-token-exchange[external to internal token exchange] is implemented only by the 

Reviewer 2: ```suggestion
* It is possible to have both the standard and the legacy token exchange features enabled, which can be useful if you need to cover standard use cases (internal-internal) together with the other token exchange use cases that h are implemented only by legacy token exchange. For instance, link:{securing_apps_token_exchange_link}#_external-token-to-internal-token-exchange[external to internal token exchange] is implemented only by the 
```

> legacy token exchange. In this case {project_name} will serve standard internal-to-internal requests preferably by the standard token exchange when the other requests will be served by the legacy token exchange. It is detected based on the

Reviewer 2: ```suggestion
legacy token exchange. In this case, {project_name} serves the standard internal-to-internal requests preferably by the standard token exchange while the other requests are served by the legacy token exchange. The method  is determined based on the
```

> parameters if the particular request is standard or not. For instance requests containing non-standard parameters like `requested_issuer` or `requested_subject` are considered legacy.

Reviewer 2: ```suggestion
parameters if the particular request is standard or not. For example, requests containing non-standard parameters such as `requested_issuer` or `requested_subject` are considered legacy.
```

> * Standard token exchange requires to enable switch on the client as described in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-enable[{securing_apps_token_exchange_name}].

Reviewer 3: ```suggestion
* Standard token exchange requires enabling a switch on the client as described in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-enable[{securing_apps_token_exchange_name}].
```

> Few more clarifications regarding the notable changes in the behaviour of standard token exchange in comparison to legacy token exchange:

Reviewer 3: ```suggestion
A few more clarifications regarding the notable changes in the behavior of the standard token exchange compared to the legacy token exchange:
```

> A few more clarifications regarding the notable changes in the behavior of the standard token exchange compared to the legacy token exchange:

Reviewer 2: ```suggestion
Consider these additional changes in the behavior of the two types of token exchange:
```

> * Fine-grained admin permissions are no longer needed and no longer supported for the standard token exchange.

Reviewer 3: ```suggestion
* Fine-grained admin permissions are no longer needed or supported for the standard token exchange.
```

> * The most notable change regarding the behaviour of scopes and audiences is, that applied client scopes are based on the client triggering the token exchange request, rather than "target" client specified by the `audience` parameter.

Reviewer 3: ```suggestion
* The most notable change regarding the behavior of scopes and audiences is that the applied client scopes are based on the client triggering the token exchange request rather than the "target" client specified by the `audience` parameter.
```

> There is support for multiple values of the `audience` parameter as mentioned in the specification. The details are described in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-scope[{securing_apps_token_exchange_name}].

Reviewer 2: ```suggestion
Support exists for multiple values of the `audience` parameter as mentioned in the specification. The details are described in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-scope[{securing_apps_token_exchange_name}].
```

> * Public clients are no longer allowed to send the token exchange requests. Legacy token exchange implemented public clients to exchanging to itself to downscope the original token. This use case can be

Reviewer 3: ```suggestion
* Public clients are no longer allowed to send the token exchange requests. Legacy token exchange allowed public clients to exchange tokens with themselves to downscope the original token. This use case can be
```

> * Public clients are no longer allowed to send the token exchange requests. Legacy token exchange allowed public clients to exchange tokens with themselves to downscope the original token. This use case can be

Reviewer 2: ```suggestion
* Public clients are no longer allowed to send the token exchange requests. Legacy token exchange allowed public clients to exchange tokens with themselves to downscope the original token. This use case can
```

> instead covered by using refresh token grant as in the refresh token grant, the parameter `scope` can be actually used to downscope the refreshed access token as mentioned in

Reviewer 3: ```suggestion
instead be covered by using the refresh token grant, in which the `scope` parameter can be used to downscope the refreshed access token, as mentioned in
```

the https://datatracker.ietf.org/doc/html/rfc6749#section-6[OAuth2 specification].

> * Exchanging an access token for a SAML assertion is not supported for now. In other words, using `requested_token_type=urn:ietf:params:oauth:token-type:saml2` is not supported.

Reviewer 2: ```suggestion
* Exchanging an access token for a SAML assertion is not supported in this release. In other words, using `requested_token_type=urn:ietf:params:oauth:token-type:saml2` is not supported.
```

> * Exchanging access token to SAML assertion is not supported for now. In other words, it is not supported to use `requested_token_type=urn:ietf:params:oauth:token-type:saml2`.

Reviewer 3: ```suggestion
* Exchanging an access token for a SAML assertion is not supported for now. In other words, using `requested_token_type=urn:ietf:params:oauth:token-type:saml2` is not supported.
```

> * Exchanging access token to refresh token is allowed just if explicitly enabled on the client as mentioned in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-details[{securing_apps_token_exchange_name}].

Reviewer 3: ```suggestion
* Exchanging an access token for a refresh token is allowed only if it is explicitly enabled on the client as mentioned in the link:{securing_apps_token_exchange_link}#_standard-token-exchange-details[{securing_apps_token_exchange_name}].
```

> Right now, it is not supported to request offline tokens or exchange refresh token when the subject token was issued from the offline session. It is recommended to exchange to access tokens instead of

Reviewer 3: ```suggestion
Currently, it is not supported to request offline tokens or exchange a refresh token when the subject token was issued from an offline session. It is recommended to exchange for access tokens instead of
```

> Currently, it is not supported to request offline tokens or exchange a refresh token when the subject token was issued from an offline session. It is recommended to exchange for access tokens instead of

Reviewer 2: ```suggestion
Currently, it is not supported to request offline tokens or exchange a refresh token when the subject token was issued from an offline session. The recommended approach is to exchange for access tokens instead of
```
